tbot_contrib/locking.py

```diff
@@ -0,0 +1,216 @@
+""" Machine Locking implementations for tbot """
+
+import abc
+import contextlib
+import typing
+import time
+import tbot
+from tbot.machine import machine, linux
+
+
+class LockManagerBase(abc.ABC):
+    """
+    Defines the "interface" that each locking implementation needs to
+    implement.
+    """
+
+    @abc.abstractmethod
+    def request_machine_lock(self, properties: dict) -> bool:
```

Comment:
The reason I used a dictionary as parameter, is that I did not want the derived class to dictate the signature of the parent, as I could imagine other LockManagers wanting other parameters, and these not being relevant to them. Is there perhaps a better way as to not have to specify all the detailed parameters in the parent than a dictionary? If this is not an issue from your view, I will happily change things as you described.

Reply:
I think it is not a good idea to allow too much flexibility in the parameters, because it will create pitfalls in terms of compatibility.  Ideally the `MachineLock` should not care about the lock-manager in use at all so that you can reuse a board config with any lab-config no matter which lock-manager it uses.

Thus I'd say it is actually a _feature_ to restrict the parameters which can be passed.

However, you're completely right that e.g. `checkpid` is not really a generic parameter, it really only applies to the one `LockManager` implementation which we have now.  `expiry` on the other hand is generic, I'd say; it is a setting which each lock-manager should provide.

I think the best solution would be to just move `checkpid` into the `LockManager` entirely and not allow `MachineLock` to customize it.  I.e. the `LockManager` gets a new `lock_checkpid` configuration option to configure whether pid-checking should be enabled on this lab-host and anything `checkpid` related is dropped from the `MachineLock`.  What do you think?

---

For completeness:  If you want to allow customization of the signature, you could design the "interface class" like this:

```python
from typing import Optional, Any

class LockManagerBase(abc.ABC):
    @abc.abstractmethod
    def request_machine_lock(
        self,
        name: str,
        *,
        expiry: Optional[int] = None,
        **kwargs: Any,
    ) -> bool:
        ...
```

This `kwargs` "catches" all additional keyword arguments and (in a function with a body) would be a dictionary with their values.

Though as written above, I don't think this is the right approach here.

Reply:
I am actually happy with moving the checkpid to LockManager, as it leaves less room for confusion. I guess if someone later requires additional arguments for another LockManager, we can do what you suggested here, but not definitely not for checkpid! Thanks for this advice.